### PR TITLE
AA-378 Completing native tracer

### PR DESCRIPTION
**Pull request by shahafn**

### CODE REVIEW

File: core/types/tx_rip7560.go [hunk 1]
```diff
@@ -187,19 +190,32 @@ func (tx *Rip7560AccountAbstractionTx) AbiEncode() ([]byte, error) {
...
 	}
+
+	paymaster := tx.Paymaster
+	if paymaster == nil {
+		paymaster = &common.Address{}
+	}
+	deployer := tx.Deployer
+	if deployer == nil {
+		deployer = &common.Address{}
+	}
```
Comment (forshtat) [inline core/types/tx_rip7560.go:201]: @shahafn @drortirosh I think we need to decide whether we treat zero address as a `nil` or if we want to actually prevent that. What do you think?

File: eth/tracers/native/rip7560_validation.go [hunk 2]
```diff
@@ -271,34 +298,44 @@ func (b *rip7560ValidationTracer) OnOpcode(pc uint64, op byte, gas, cost uint64,
 			if !rOk && !wOk {
 				access.Reads[slotHex] = b.env.StateDB.GetState(addr, slot).Hex()
 			}
-		} else {
+		} else if opcode == "STORE" {
```
Comment (forshtat) [inline eth/tracers/native/rip7560_validation.go:301]: The opcode is called `SSTORE` not `STORE` so I assume this "if" will never be entered. Is it covered by a test?

**Review by forshtat — APPROVED** [review]

File: core/types/tx_rip7560.go (hunk 1 above)
Comment (shahafn) [reply core/types/tx_rip7560.go:201]: Currently it crashes when abi encoding it with nil, so we need to put zero values there

File: eth/tracers/native/rip7560_validation.go [hunk 3]
```diff
@@ -142,6 +146,27 @@ func (b *rip7560ValidationTracer) OnEnter(depth int, typ byte, from common.Addre
...
+		Data:  input,
+	})
+}
+
+func (b *rip7560ValidationTracer) OnExit(depth int, output []byte, gasUsed uint64, err error, reverted bool) {
+	typ := "RETURN"
+	if err != nil {
+		typ = "REVERT"
+	}
+	b.Calls = append(b.Calls, &callsItem{
```
[the remark this answers is not in the document]
Comment (shahafn) [reply eth/tracers/native/rip7560_validation.go:165]: We don't create the same entries, and here the calls with values are properly passed

File: eth/tracers/native/rip7560_validation.go (hunk 2 above)
Comment (shahafn) [reply eth/tracers/native/rip7560_validation.go]: Yes, fixed and added test cases specific to SSTORE

**Review by forshtat — APPROVED** [review]

**Pull request merged**